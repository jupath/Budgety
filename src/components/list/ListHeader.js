import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

export const ListHeader = props => (
  <div className="list__header py-3 px-2">
    <div>
      <h4 className="mb-0">{props.listName}</h4>
    </div>
    <div>
      <Link className="btn btn-info" to={`/add/${props.itemsType}`}>Add new {props.itemsType}</Link>
    </div>
  </div>
);

ListHeader.propTypes = {
  itemsType: PropTypes.string.isRequired,
  listName: PropTypes.string.isRequired,
};

export default ListHeader;
