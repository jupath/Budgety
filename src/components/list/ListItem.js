import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

const ListItem = props => (
  <Link to={`/edit/${props.listItem.id}`}>
    <div className="list__item">
      <div>
        {props.listItem.description}<br />
        {moment(props.listItem.date).format('MMMM Do, YYYY')}
      </div>
      <div>
        {props.listItem.amount}
      </div>
    </div>
  </Link>
);

ListItem.propTypes = {
  listItem: PropTypes.shape({
    itemType: PropTypes.string,
    description: PropTypes.string,
    amount: PropTypes.number,
    date: PropTypes.number,
    note: PropTypes.string,
    id: PropTypes.string,
  }).isRequired,
};

export default ListItem;
