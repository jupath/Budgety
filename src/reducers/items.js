export default (state = [], action) => {
  switch (action.type) {
    case 'ADD_ITEM':
      return [...state, action.item];
    case 'UPDATE_ITEM':
      return state.map((item) => {
        if (item.id === action.id) {
          return {
            ...item,
            ...action.item,
          };
        }
        return item;
      });
    case 'DELETE_ITEM':
      return state.filter(item => item.id !== action.id);
    default:
      return state;
  }
};
