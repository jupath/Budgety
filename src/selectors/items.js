import moment from 'moment';

export default (items, itemsType, { startDate, endDate, sortBy }) => items.filter((item) => {
  const date = moment(item.date);
  const startDateMatch = startDate ? startDate.isSameOrBefore(date, 'day') : true;
  const endDateMatch = endDate ? endDate.isSameOrAfter(date, 'day') : true;
  const itemTypeMatch = item.itemType === itemsType;

  return startDateMatch && endDateMatch && itemTypeMatch;
}).sort((a, b) => {
  if (sortBy === 'date') {
    return a.date < b.date ? 1 : -1;
  } else if (sortBy === 'value') {
    return a.amount < b.amount ? 1 : -1;
  }
});
