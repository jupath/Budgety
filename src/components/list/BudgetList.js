import React from 'react';
import PropTypes from 'prop-types';
import ListItem from './ListItem';

export const BudgetList = props => (
  <div>
    {
      props.items.map(item => <ListItem key={item.id} listItem={item} />)
    }
  </div>
);

BudgetList.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default BudgetList;
