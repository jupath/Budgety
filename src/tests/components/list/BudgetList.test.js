import React from 'react';
import { shallow } from 'enzyme';
import { BudgetList } from '../../../components/list/BudgetList';
import items from '../../fixtures/items';

test('should render BudgetList correctly', () => {
  const wrapper = shallow(<BudgetList items={items} />);
  expect(wrapper).toMatchSnapshot();
});
