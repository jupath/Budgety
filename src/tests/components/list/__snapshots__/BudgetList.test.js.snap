// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BudgetList correctly 1`] = `
<div>
  <ListItem
    key="abc123"
    listItem={
      Object {
        "amount": 3100,
        "date": 0,
        "description": "Test description 1",
        "id": "abc123",
        "itemType": "expense",
        "note": "",
      }
    }
  />
  <ListItem
    key="bcd234"
    listItem={
      Object {
        "amount": 5000.52,
        "date": -345600000,
        "description": "Test description 2",
        "id": "bcd234",
        "itemType": "income",
        "note": "Salery",
      }
    }
  />
  <ListItem
    key="def345"
    listItem={
      Object {
        "amount": 2000,
        "date": 345600000,
        "description": "Test description 3",
        "id": "def345",
        "itemType": "expense",
        "note": "Dine out",
      }
    }
  />
</div>
`;
